### PR TITLE
address cmd_exec inconsistencies with mettle/meterpreter payloads

### LINKED ISSUE #9429: `cmd_exec` inconsistent behaviour between Meterpreter and Shell sessions
`cmd_exec` is handled differently in shell and Meterpreter sessions.

## Steps to reproduce

Test module `modules/exploits/linux/local/test.rb` :

```ruby
##
# This module requires Metasploit: https://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##

class MetasploitModule < Msf::Exploit::Local
  Rank = ExcellentRanking

  include Msf::Post::File
  include Msf::Exploit::EXE
  include Msf::Exploit::FileDropper

  def initialize(info = {})
    super(update_info(info,
      'Name'           => 'Test',
      'Description'    => %q{
        Test
      },
      'License'        => MSF_LICENSE,
      'Author'         =>
        [
          'test'
        ],
      'DisclosureDate' => 'test',
      'Platform'       => [ 'linux' ],
      'Arch'           => [ ARCH_X86, ARCH_X64 ],
      'SessionTypes'   => [ 'shell', 'meterpreter' ],
      'Targets'        => [[ 'Auto', {} ]],
      'References'     =>
        [
          [ ]
        ]
    ))
    register_options(
      [
        OptInt.new('TIMEOUT', [ true, 'Timeout (seconds)', '60' ])
      ])
  end

  def timeout
    datastore['TIMEOUT']
  end

  def exploit
    cmd = '/bin/ls -l /etc/passwd'
    args = nil
    print_status "Running command: #{cmd.inspect}"
    output = cmd_exec cmd, args, timeout
    output.each_line { |line| vprint_status line.chomp }
    puts 'command completed successfully' if output =~ /root/

    puts cmd_exec 'echo we made it to the end'
  end
end
```

## Output

```
msf exploit(linux/local/test) > sessions 

Active sessions
===============

  Id  Name  Type                   Information                                                       Connection
  --  ----  ----                   -----------                                                       ----------
  1         shell cmd/unix                                                                           172.16.191.244:1337 -> 172.16.191.137:37369 (172.16.191.137)
  6         meterpreter x86/linux  uid=1000, gid=1000, euid=1000, egid=1000 @ localhost.localdomain  172.16.191.244:4433 -> 172.16.191.137:49768 (172.16.191.137)
```

### Shell Session

```
set sessimsf exploit(linux/local/test) > set session 1
session => 1
msf exploit(linux/local/test) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Running command: "/bin/ls -l /etc/passwd"
[*] -rw-r--r--. 1 root root 2076 Jan 18 02:59 /etc/passwd
command completed successfully
we made it to the end
[*] Exploit completed, but no session was created.
```

### Meterpreter Session

```
msf exploit(linux/local/test) > set session 6
session => 6
msf exploit(linux/local/test) > run

[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Running command: "/bin/ls -l /etc/passwd"
[*] (current) UNIX password:        # <-- wtf
we made it to the end
[*] Exploit completed, but no session was created.
```

## Expected behavior

`puts 'command completed successfully'` for both shell and Meterpreter sessions.

## Current behavior

`puts 'command completed successfully'` for only shell sessions.

## Notes

It's possible that `def cmd_exec(cmd, args=nil, time_out=15)` was intended to be used in the form of `'/path/to/command', 'argument', int_timeout` - but that's no excuse for inconsistent and undocumented behavior dependent on session type.

Additionally, using the above method invocation introduced unrelated issues (again, only with Meterpreter sessions and not shell sessions) when dealing with `timeout` handling.

Eyeballing the `cmd_exec` method in `lib/msf/core/post/common.rb` (shown below) shows that `break` is invoked inside the `while (d = process.channel.read)` loop if execution time has reached `timeout` or we received data during a previous iteration but not this iteration. That makes no sense. That's bad.

```ruby
  def cmd_exec(cmd, args=nil, time_out=15)
    case session.type
    when /meterpreter/

      # --- removed for brevity ---

      session.response_timeout = time_out
      process = session.sys.process.execute(cmd, args, {'Hidden' => true, 'Channelized' => true})
      o = ""
      # Wait up to time_out seconds for the first bytes to arrive
      while (d = process.channel.read)
        if d == ""
          if (Time.now.to_i - start < time_out) && (o == '')
            sleep 0.1
          else
            break
          end
        else
          o << d
        end
      end
      o.chomp! if o
```

### PR DESCRIPTION
**Pull request by busterb**
Fixes #9429 

This attempts to address the inconsistent behavior. Test cases are in the original issue.

@bcoles would you mind giving this a quick pass? Thanks!

This took a little while to push up, had some interesting CI corner cases that had to be fixed first. This should also fix the crash observed when running the new 'rename process' functionality from an injected process.

### CODE REVIEW

**Comment by bcoles** [comment]
Changes appears to work as described.

* ✔️ Segfault: fixed 
* ✔️ Timeout: fixed 
* ✔️ `cmd_exec cmd` style invocation: fixed 

I'll assume your `CachedSize` changes are sane (I love inexplicable unsigned integers as much as the next person, but perhaps there should be a constant for each payload platform defined somewhere ?).

Output, with Fedora 20 x64 test client:

```
msf5 > use exploit/multi/handler 
msf5 exploit(multi/handler) > set lport 1337
lport => 1337
msf5 exploit(multi/handler) > set lhost 172.16.191.244
lhost => 172.16.191.244
msf5 exploit(multi/handler) > set payload cmd/unix/reverse_netcat
payload => cmd/unix/reverse_netcat
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 172.16.191.244:1337 
[*] Command shell session 1 opened (172.16.191.244:1337 -> 172.16.191.137:37488) at 2018-01-21 17:20:39 -0500


id
uid=1000(user) gid=1000(user) groups=1000(user),10(wheel) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
^Z
Background session 1? [y/N]  y

msf5 exploit(multi/handler) > use exploit/linux/local/test
msf5 exploit(linux/local/test) > info

       Name: Test
     Module: exploit/linux/local/test
   Platform: Linux
       Arch: x86, x64
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent

Provided by:
  test

Available targets:
  Id  Name
  --  ----
  0   Auto

Basic options:
  Name     Current Setting  Required  Description
  ----     ---------------  --------  -----------
  SESSION                   yes       The session to run this module on.
  TIMEOUT  60               yes       Timeout (seconds)

Payload information:

Description:
  Test

msf5 exploit(linux/local/test) > run

[-] Exploit failed: The following options failed to validate: SESSION.
[*] Exploit completed, but no session was created.
msf5 exploit(linux/local/test) > set session 1
session => 1
msf5 exploit(linux/local/test) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Running command: "/bin/ls -l /etc/passwd"
command completed successfully
we made it to the end


[*] Exploit completed, but no session was created.

msf5 exploit(linux/local/test) > 

msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.191.244:4433 
[*] Sending stage (857352 bytes) to 172.16.191.137
[*] Meterpreter session 2 opened (172.16.191.244:4433 -> 172.16.191.137:49883) at 2018-01-21 17:21:09 -0500
[*] Command stager progress: 100.00% (773/773 bytes)
msf5 exploit(linux/local/test) > sessions -k
[-] Please specify valid session identifier(s)
msf5 exploit(linux/local/test) > sessions -l

Active sessions
===============

  Id  Name  Type                   Information                                                       Connection
  --  ----  ----                   -----------                                                       ----------
  1         shell cmd/unix                                                                           172.16.191.244:1337 -> 172.16.191.137:37488 (172.16.191.137)
  2         meterpreter x86/linux  uid=1000, gid=1000, euid=1000, egid=1000 @ localhost.localdomain  172.16.191.244:4433 -> 172.16.191.137:49883 (172.16.191.137)

msf5 exploit(linux/local/test) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > getuid
Server username: uid=1000, gid=1000, euid=1000, egid=1000
meterpreter > 
meterpreter > 
meterpreter > 
Background session 2? [y/N]  
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > set session 2
session => 2
msf5 exploit(linux/local/test) > run

[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Running command: "/bin/ls -l /etc/passwd"
command completed successfully
we made it to the end
[*] Exploit completed, but no session was created.
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > 
msf5 exploit(linux/local/test) > use exploit/linux/local/test2
msf5 exploit(linux/local/test2) > set session 1
session => 1
rmsf5 exploit(linux/local/test2) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Running command: "/bin/echo" with args "some_initial_data && sleep 20 && echo true"
command completed successfully
we made it to the end
[*] Exploit completed, but no session was created.
msf5 exploit(linux/local/test2) > set session 2
session => 2
msf5 exploit(linux/local/test2) > run

[*] Started reverse TCP handler on 172.16.191.244:4444 
[*] Running command: "/bin/echo" with args "some_initial_data && sleep 20 && echo true"
command completed successfully
we made it to the end
[*] Exploit completed, but no session was created.
```

File: lib/msf/core/post/common.rb [hunk 1]
```diff
@@ -111,14 +111,13 @@ def cmd_exec(cmd, args=nil, time_out=15)
       o = ""
       # Wait up to time_out seconds for the first bytes to arrive
       while (d = process.channel.read)
+        o << d
         if d == ""
-          if (Time.now.to_i - start < time_out) && (o == '')
```
Comment (sempervictus) [inline lib/msf/core/post/common.rb]: Seriously? Thank you...

**Comment by busterb** [comment]
Thanks for testing @bcoles , now on to the PHP payload :P

The CachedSize integers are generated by a script, and for legacy reasons, our spec tests are sticklers about those. I put those changes in a separate commit to make review easier. Let me noodle on a better way we might be able to solve the problem CachedSize solves, because I do feel like we're wasting time valuable time as humans every time we update them.

**Review by sempervictus — APPROVED** [review]
Sometimes a rational fix devolves into a game of whackamole to find all the workarounds previously implemented... Thanks.

**Comment by busterb** [comment]
## Release Notes

This fixes issues with the cmd_exec command behaving inconsistently with Meterpreter versus shell payloads. It also fixes some reverse_http issues when reconnecting via Linux/OSX native Meterpreters.
